File: python/examples/hello_world_pub.py
#!/usr/bin/env python3
"""
HelloWorld Publisher Example

Publishes HelloWorld samples to demonstrate int2dds Python bindings.

Usage:
    python hello_world_pub.py [-d DOMAIN] [--reliable]
"""

import argparse
import time
import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from hello_world_type import HelloWorld

from int2dds import DomainParticipant, WaitSet, DataWriterQos, Reliability

TOPIC_NAME = "hello_world_topic"


def _kind_name(kind: str) -> str:
    """Render a QoS kind the way the Rust example's Debug output does."""
    return "".join(part.capitalize() for part in kind.split("_"))


def _history_name(history) -> str:
    if history.kind == "KEEP_ALL":
        return "KeepAll"
    return f"KeepLast({history.depth})"


def main() -> None:
    # Reliability and domain are selectable on the CLI, matching the Rust/C#/C examples.
    parser = argparse.ArgumentParser(description="HelloWorld DDS Publisher (Python)")
    parser.add_argument("-d", "--domain", type=int, default=0, help="Domain ID (default 0)")
    parser.add_argument("--reliable", action="store_true",
                        help="Use RELIABLE reliability (default BEST_EFFORT)")
    args = parser.parse_args()

    # Create domain participant
    with DomainParticipant(domain_id=args.domain, name="PythonPublisher") as dp:
        # Create topic
        topic = dp.create_topic(TOPIC_NAME, HelloWorld)

        # Create publisher and data writer (BEST_EFFORT by default, --reliable for RELIABLE)
        pub = dp.create_publisher()
        reliability = (
            Reliability("RELIABLE", max_blocking_time=0.1)
            if args.reliable
            else Reliability("BEST_EFFORT")
        )
        writer = pub.create_datawriter(topic, DataWriterQos(reliability=reliability))

        wqos = writer.get_qos()
        print(f"[publisher INFO] domain_id: {args.domain}, topic: {TOPIC_NAME}")
        print(
            f"[publisher qos] reliability: {_kind_name(wqos.reliability.kind)}, "
            f"durability: {_kind_name(wqos.durability.kind)}, "
            f"history: {_history_name(wqos.history)}"
        )

        # Wait for subscriber to connect
        waitset = WaitSet()
        waitset.attach(writer)

        # Run until Ctrl-C, then the participant context manager cleans up
        # gracefully (matches the Rust example).
        try:
            while writer.matched_readers == 0:
                try:
                    waitset.wait(timeout=1.0)
                except Exception:
                    pass  # Timeout, check again

            print("Subscriber matched!")

            # Publish samples until Ctrl-C, like the Rust/C#/C examples
            i = 1
            while True:
                sample = HelloWorld(index=i, message=f"[Python]HelloWorld_d{args.domain}")
                writer.write(sample)
                print(f'Published HelloWorld {{ index: {sample.index}, message: "{sample.message}" }}')
                time.sleep(1.0)
                i += 1
        except KeyboardInterrupt:
            print("\nShutting down...")


if __name__ == "__main__":
    main()

File: java/examples/src/main/java/com/intellectus/int2dds/examples/HelloWorldPub.java
package com.intellectus.int2dds.examples;

import com.intellectus.int2dds.core.DataWriter;
import com.intellectus.int2dds.core.DomainParticipant;
import com.intellectus.int2dds.core.Publisher;
import com.intellectus.int2dds.core.Topic;
import com.intellectus.int2dds.qos.DataWriterQos;
import com.intellectus.int2dds.qos.Reliability;
import com.intellectus.int2dds.qos.ReliabilityKind;

/**
 * Publishes a {@link HelloWorld} sample once a second.
 *
 * <p>Modeled on the C# {@code HelloWorldPub} example, with one deliberate
 * difference: it does not wait for a subscriber before publishing. The C#
 * example attaches a {@code StatusCondition} to a {@code WaitSet} and blocks
 * until {@code PublicationMatched} fires; conditions are not part of this
 * branch yet. This example simply publishes on a fixed interval and prints
 * what it sent — that is honest about what a write-only branch can prove: it
 * confirms the write path accepted the sample, not that any subscriber
 * received it.
 *
 * <p>Run with {@code -d}/{@code --domain <id>} (default 0, the real DDS
 * default domain) and {@code --reliable} (default BEST_EFFORT), the same
 * flags the C# and Rust examples take.
 */
public final class HelloWorldPub {

    private static final String TOPIC_NAME = "hello_world_topic";
    private static final long PUBLISH_INTERVAL_MS = 1000L;

    public static void main(String[] args) throws InterruptedException {
        int domainId = parseDomain(args);
        boolean reliable = hasFlag(args, "--reliable");

        System.out.println("=== HelloWorld Publisher (Java) ===");
        System.out.println("QoS: " + (reliable ? "RELIABLE" : "BEST_EFFORT"));

        // A single try-with-resources on the participant: NativeEntity.close()
        // cascades to every live child (topic, publisher, writer), so an
        // exception while any of them is being built still releases whatever
        // was already created -- nothing below needs its own try block.
        try (DomainParticipant participant = new DomainParticipant(domainId)) {
            // The loop below runs until the process is killed (Ctrl-C, or the
            // timeout used to bound this example in CI); that never returns
            // from the try block normally, so this hook -- not the
            // try-with-resources -- is what actually releases the participant
            // on the way out.
            Runtime.getRuntime().addShutdownHook(new Thread(participant::close));
            System.out.println("Created participant on domain " + participant.domainId());

            Topic<HelloWorld> topic = participant.createTopic(TOPIC_NAME, new HelloWorld());
            System.out.println("Created topic: " + topic.name() + " (" + topic.typeName() + ")");

            Publisher publisher = participant.createPublisher();
            ReliabilityKind kind = reliable ? ReliabilityKind.RELIABLE : ReliabilityKind.BEST_EFFORT;
            DataWriterQos qos = new DataWriterQos();
            qos.setReliability(new Reliability(kind));
            DataWriter<HelloWorld> writer = publisher.createDataWriter(topic, qos);
            System.out.println("Created publisher and data writer");

            int index = 0;
            while (true) {
                HelloWorld sample = new HelloWorld();
                sample.index = index;
                sample.message = "Hello from Java! (" + index + ")";
                writer.write(sample);
                // Publishing succeeds whether or not a subscriber is listening --
                // this line only says the sample reached the write path, not that
                // anyone received it. See the class doc.
                System.out.println(
                        "Published: index=" + sample.index + ", message='" + sample.message + "'");
                Thread.sleep(PUBLISH_INTERVAL_MS);
                index++;
            }
        }
    }

    /** Parses {@code -d}/{@code --domain <id>} from the CLI, defaulting to 0. */
    private static int parseDomain(String[] args) {
        for (int i = 0; i < args.length - 1; i++) {
            if (args[i].equals("-d") || args[i].equals("--domain")) {
                return Integer.parseInt(args[i + 1]);
            }
        }
        return 0;
    }

    private static boolean hasFlag(String[] args, String flag) {
        for (String arg : args) {
            if (arg.equals(flag)) {
                return true;
            }
        }
        return false;
    }
}
